### ISSUE TITLE
TEST-55-OOMD is flaky in Github Actions

### ISSUE THREAD

**Issue by daandemeyer** [systemd/systemd#32730]
### systemd version the issue has been seen with

main

### CPU architectures issue was seen on

None

### Unexpected behaviour you saw

https://github.com/systemd/systemd/actions/runs/9005391896/job/24740590109

```

gh run download 9005391896 --name ci-mkosi-9005391896-1-fedora-rawhide-failed-test-journals -D ci/ci-mkosi-9005391896-1-fedora-rawhide-failed-test-journals && journalctl --file ci/ci-mkosi-9005391896-1-fedora-rawhide-failed-test-journals/test/journal/TEST-55-OOMD.journal --no-hostname -o short-monotonic -u testsuite-55.service -p info
```

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
34/57 systemd:integration-tests / TEST-55-OOMD                  FAIL             57.28s   exit status 24
>>> MALLOC_PERTURB_=223 SYSTEMD_INTEGRATION_TESTS=1 /home/runner/work/systemd/systemd/test/integration-test-wrapper.py --meson-source-dir /home/runner/work/systemd/systemd --meson-build-dir /mnt/mkosi/build --test-name TEST-55-OOMD --test-number 55 --storage volatile --firmware linux -- --credential=/home/runner/work/systemd/systemd/test/TEST-55-OOMD/systemd.unit-dropin.init.scope
――――――――――――――――――――――――――――――――――――― ✀  ―――――――――――――――――――――――――――――――――――――
Listing only the last 100 lines from a long log.
[   47.836933] testsuite-55.sh[924]: 		Memory Min: 0B
[   47.838354] testsuite-55.sh[924]: 		Memory Low: 0B
[   47.839750] testsuite-55.sh[924]: 		Pgscan: 45631
[   47.841746] testsuite-55.sh[924]: 		Last Pgscan: 45603
[   47.843269] testsuite-55.sh[924]: 	Path: /user.slice/user-4711.slice/user@4711.service/testsuite.slice/testsuite-55.slice/testsuite-55-workload.slice
[   47.846860] testsuite-55.sh[924]: 		Memory Pressure Limit: 80.00%
[   47.848707] testsuite-55.sh[924]: 		Pressure: Avg10: 81.26 Avg60: 24.68 Avg300: 5.68 Total: 17s
[   47.851214] testsuite-55.sh[924]: 		Current Memory Usage: 5.7M
[   47.853069] testsuite-55.sh[924]: 		Memory Min: 0B
[   47.854499] testsuite-55.sh[924]: 		Memory Low: 0B
[   47.855925] testsuite-55.sh[924]: 		Pgscan: 45631
[   47.857374] testsuite-55.sh[924]: 		Last Pgscan: 45603
[   47.858896] testsuite-55.sh[924]: 	Path: /user.slice/user-4711.slice/user@4711.service/testsuite.slice
[   47.861443] testsuite-55.sh[924]: 		Memory Pressure Limit: 80.00%
[   47.863460] testsuite-55.sh[924]: 		Pressure: Avg10: 81.22 Avg60: 24.66 Avg300: 5.68 Total: 17s
[   47.865931] testsuite-55.sh[924]: 		Current Memory Usage: 5.8M
[   47.871481] testsuite-55.sh[924]: 		Memory Min: 0B
[   47.877884] testsuite-55.sh[924]: 		Memory Low: 0B
[   47.879616] testsuite-55.sh[924]: 		Pgscan: 45631
[   47.884929] testsuite-55.sh[924]: 		Last Pgscan: 45603
[   47.887392] testsuite-55.sh[924]: 	Path: /user.slice/user-4711.slice/user@4711.service/session.slice
[   47.899640] testsuite-55.sh[924]: 		Memory Pressure Limit: 80.00%
[   47.901836] testsuite-55.sh[924]: 		Pressure: Avg10: 0.00 Avg60: 0.00 Avg300: 0.00 Total: 0
[   47.908112] testsuite-55.sh[924]: 		Current Memory Usage: 820.0K
[   47.916138] testsuite-55.sh[924]: 		Memory Min: 0B
[   47.918921] testsuite-55.sh[924]: 		Memory Low: 0B
[   47.920261] testsuite-55.sh[924]: 		Pgscan: 0
[   47.923456] testsuite-55.sh[924]: 		Last Pgscan: 0
[   47.925196] testsuite-55.sh[924]: 	Path: /user.slice/user-4711.slice/user@4711.service/app.slice
[   47.927894] testsuite-55.sh[924]: 		Memory Pressure Limit: 80.00%
[   47.934058] testsuite-55.sh[924]: 		Pressure: Avg10: 0.00 Avg60: 0.00 Avg300: 0.00 Total: 0
[   47.936203] testsuite-55.sh[924]: 		Current Memory Usage: 16.0K
[   47.937874] testsuite-55.sh[924]: 		Memory Min: 0B
[   47.939868] testsuite-55.sh[924]: 		Memory Low: 0B
[   47.941523] testsuite-55.sh[924]: 		Pgscan: 0
[   47.943217] testsuite-55.sh[924]: 		Last Pgscan: 0
[   47.946133] testsuite-55.sh[924]: 	Path: /testsuite.slice/testsuite-55.slice/testsuite-55-workload.slice
[   47.950519] testsuite-55.sh[924]: 		Memory Pressure Limit: 20.00%
[   47.955319] testsuite-55.sh[924]: 		Pressure: Avg10: 6.76 Avg60: 7.11 Avg300: 1.96 Total: 6s
[   47.957859] testsuite-55.sh[924]: 		Current Memory Usage: 256.0K
[   47.959726] testsuite-55.sh[924]: 		Memory Min: 0B
[   47.961203] testsuite-55.sh[924]: 		Memory Low: 0B
[   47.962680] testsuite-55.sh[924]: 		Pgscan: 44889
[   47.964383] testsuite-55.sh[924]: 		Last Pgscan: 44889
[   47.965922] testsuite-55.sh[924]: 	Path: /user.slice/user-4711.slice/user@4711.service
[   47.968337] testsuite-55.sh[924]: 		Memory Pressure Limit: 80.00%
[   47.971171] testsuite-55.sh[924]: 		Pressure: Avg10: 81.18 Avg60: 24.63 Avg300: 5.67 Total: 17s
[   47.973328] testsuite-55.sh[924]: 		Current Memory Usage: 13.3M
[   47.975105] testsuite-55.sh[924]: 		Memory Min: 0B
[   47.977433] testsuite-55.sh[924]: 		Memory Low: 0B
[   47.978664] testsuite-55.sh[924]: 		Pgscan: 45631
[   47.979857] testsuite-55.sh[924]: 		Last Pgscan: 45603
[   47.981163] testsuite-55.sh[924]: 	Path: /user.slice/user-0.slice/user@0.service/app.slice
[   47.983485] testsuite-55.sh[924]: 		Memory Pressure Limit: 80.00%
[   47.985249] testsuite-55.sh[924]: 		Pressure: Avg10: 0.00 Avg60: 0.00 Avg300: 0.00 Total: 0
[   47.987266] testsuite-55.sh[924]: 		Current Memory Usage: 12.0K
[   47.988782] testsuite-55.sh[924]: 		Memory Min: 0B
[   47.989979] testsuite-55.sh[924]: 		Memory Low: 0B
[   47.993668] testsuite-55.sh[924]: 		Pgscan: 0
[   47.995630] testsuite-55.sh[924]: 		Last Pgscan: 0
[   47.997567] testsuite-55.sh[588]: + sleep 2
[   49.738233] testsuite-55.sh[588]: + for _ in {0..59}
[   49.739894] testsuite-55.sh[588]: + systemctl --machine testuser@.host --user status testsuite-55-testbloat.service
[   49.831489] testsuite-55.sh[926]: × testsuite-55-testbloat.service - Create a lot of memory pressure
[   49.834380] testsuite-55.sh[926]:      Loaded: loaded (/usr/lib/systemd/tests/testdata/units/testsuite-55-testbloat.service; static)
[   49.838207] testsuite-55.sh[926]:     Drop-In: /usr/lib/systemd/user/service.d
[   49.841346] testsuite-55.sh[926]:              └─10-timeout-abort.conf
[   49.842822] testsuite-55.sh[926]:      Active: failed (Result: oom-kill) since Wed 2024-05-08 17:08:58 UTC; 27ms ago
[   49.844962] testsuite-55.sh[926]:    Duration: 20.595s
[   49.846306] testsuite-55.sh[926]:  Invocation: 7dca6b0bd35e43649a7f43502e79fee9
[   49.847809] testsuite-55.sh[926]:     Process: 791 ExecStart=stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1 (code=killed, signal=KILL)
[   49.850517] testsuite-55.sh[926]:    Main PID: 791 (code=killed, signal=KILL)
[   49.852373] testsuite-55.sh[588]: + break
[   49.853344] testsuite-55.sh[588]: + systemctl --machine testuser@.host --user status testsuite-55-testbloat.service
[   49.922912] testsuite-55.sh[936]: × testsuite-55-testbloat.service - Create a lot of memory pressure
[   49.929620] testsuite-55.sh[936]:      Loaded: loaded (/usr/lib/systemd/tests/testdata/units/testsuite-55-testbloat.service; static)
[   49.932273] testsuite-55.sh[936]:     Drop-In: /usr/lib/systemd/user/service.d
[   49.934706] testsuite-55.sh[936]:              └─10-timeout-abort.conf
[   49.937070] testsuite-55.sh[936]:      Active: failed (Result: oom-kill) since Wed 2024-05-08 17:08:58 UTC; 117ms ago
[   49.940389] testsuite-55.sh[936]:    Duration: 20.595s
[   49.942363] testsuite-55.sh[936]:  Invocation: 7dca6b0bd35e43649a7f43502e79fee9
[   49.944821] testsuite-55.sh[936]:     Process: 791 ExecStart=stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1 (code=killed, signal=KILL)
[   49.949092] testsuite-55.sh[936]:    Main PID: 791 (code=killed, signal=KILL)
[   49.951429] testsuite-55.sh[588]: + systemctl --machine testuser@.host --user status testsuite-55-testchill.service
[   50.040374] testsuite-55.sh[942]: × testsuite-55-testchill.service - No memory pressure
[   50.043580] testsuite-55.sh[942]:      Loaded: loaded (/usr/lib/systemd/tests/testdata/units/testsuite-55-testchill.service; static)
[   50.047687] testsuite-55.sh[942]:     Drop-In: /usr/lib/systemd/user/service.d
[   50.050494] testsuite-55.sh[942]:              └─10-timeout-abort.conf
[   50.052999] testsuite-55.sh[942]:      Active: failed (Result: oom-kill) since Wed 2024-05-08 17:08:58 UTC; 254ms ago
[   50.056374] testsuite-55.sh[942]:    Duration: 20.781s
[   50.058299] testsuite-55.sh[942]:  Invocation: 49dac7991dc34a269c53098c2003c214
[   50.061113] testsuite-55.sh[942]:     Process: 781 ExecStart=sleep infinity (code=killed, signal=KILL)
[   50.065083] testsuite-55.sh[942]:    Main PID: 781 (code=killed, signal=KILL)
[   50.068254] testsuite-55.sh[588]: + exit 24
```

**Comment by yuwata** [comment]
Another failure:
https://github.com/systemd/systemd/actions/runs/9334201663/job/25692125515?pr=33154
```
[   31.387422] TEST-55-OOMD.sh[819]: 		Memory Min: 0B
[   31.394577] TEST-55-OOMD.sh[819]: 		Memory Low: 0B
[   31.396354] TEST-55-OOMD.sh[819]: 		Pgscan: 0
[   31.405633] TEST-55-OOMD.sh[819]: 		Last Pgscan: 0
[   31.420370] TEST-55-OOMD.sh[819]: 	Path: /system.slice
[   31.421965] TEST-55-OOMD.sh[819]: 		Memory Pressure Limit: 80.00%
[   31.425679] TEST-55-OOMD.sh[819]: 		Pressure: Avg10: 0.00 Avg60: 0.00 Avg300: 0.00 Total: 330us
[   31.431085] TEST-55-OOMD.sh[819]: 		Current Memory Usage: 225.1M
[   31.444351] TEST-55-OOMD.sh[819]: 		Memory Min: 0B
[   31.446092] TEST-55-OOMD.sh[819]: 		Memory Low: 0B
[   31.447670] TEST-55-OOMD.sh[819]: 		Pgscan: 0
[   31.449262] TEST-55-OOMD.sh[819]: 		Last Pgscan: 0
[   31.450923] TEST-55-OOMD.sh[819]: 	Path: /user.slice/user-4711.slice/user@4711.service/app.slice
[   31.457451] TEST-55-OOMD.sh[819]: 		Memory Pressure Limit: 80.00%
[   31.461772] TEST-55-OOMD.sh[819]: 		Pressure: Avg10: 0.00 Avg60: 0.00 Avg300: 0.00 Total: 0
[   31.468322] TEST-55-OOMD.sh[819]: 		Current Memory Usage: 16K
[   31.473209] TEST-55-OOMD.sh[819]: 		Memory Min: 0B
[   31.477476] TEST-55-OOMD.sh[819]: 		Memory Low: 0B
[   31.478976] TEST-55-OOMD.sh[819]: 		Pgscan: 0
[   31.482731] TEST-55-OOMD.sh[819]: 		Last Pgscan: 0
[   31.489871] TEST-55-OOMD.sh[819]: 	Path: /TEST.slice/TEST-55.slice/TEST-55-OOMD.slice/TEST-55-OOMD-workload.slice
[   31.493461] TEST-55-OOMD.sh[819]: 		Memory Pressure Limit: 20.00%
[   31.500467] TEST-55-OOMD.sh[819]: 		Pressure: Avg10: 40.51 Avg60: 8.36 Avg300: 1.78 Total: 5s
[   31.506725] TEST-55-OOMD.sh[819]: 		Current Memory Usage: 5.2M
[   33.290686] TEST-55-OOMD.sh[826]:     Drop-In: /usr/lib/systemd/system/service.d
[   33.292989] TEST-55-OOMD.sh[826]:              └─10-timeout-abort.conf
[   33.295882] TEST-55-OOMD.sh[826]:              /run/systemd/system/TEST-55-OOMD-testchill.service.d
[   33.298736] TEST-55-OOMD.sh[826]:              └─99-MemoryHigh.conf
[   33.300840] TEST-55-OOMD.sh[826]:      Active: failed (Result: oom-kill) since Sun 2024-06-02 01:13:42 UTC; 1s ago
[   33.303888] TEST-55-OOMD.sh[826]:    Duration: 7.421s
[   33.305643] TEST-55-OOMD.sh[826]:  Invocation: abe604b3f1a24bb2995636c0eb734602
[   33.308165] TEST-55-OOMD.sh[826]:     Process: 743 ExecStart=sleep infinity (code=killed, signal=KILL)
[   33.311036] TEST-55-OOMD.sh[826]:    Main PID: 743 (code=killed, signal=KILL)
[   33.313233] TEST-55-OOMD.sh[826]: Jun 02 01:13:42 H systemd[1]: TEST-55-OOMD-testchill.service: systemd-oomd killed some process(es) in this unit.
[   33.317211] TEST-55-OOMD.sh[826]: Jun 02 01:13:42 H systemd[1]: TEST-55-OOMD-testchill.service: Process(es) of control group were killed by systemd-oomd.
[   33.321298] TEST-55-OOMD.sh[826]: Jun 02 01:13:42 H systemd[1]: TEST-55-OOMD-testchill.service: Changed running -> stop-sigterm
[   33.324889] TEST-55-OOMD.sh[826]: Jun 02 01:13:42 H systemd[1]: TEST-55-OOMD-testchill.service: Child 743 belongs to TEST-55-OOMD-testchill.service.
[   33.328834] TEST-55-OOMD.sh[826]: Jun 02 01:13:42 H systemd[1]: TEST-55-OOMD-testchill.service: Main process exited, code=killed, status=9/KILL
[   33.332710] TEST-55-OOMD.sh[826]: Jun 02 01:13:42 H systemd[1]: TEST-55-OOMD-testchill.service: Failed with result 'oom-kill'.
[   33.336181] TEST-55-OOMD.sh[826]: Jun 02 01:13:42 H systemd[1]: TEST-55-OOMD-testchill.service: Service will not restart (restart setting)
[   33.340222] TEST-55-OOMD.sh[826]: Jun 02 01:13:42 H systemd[1]: TEST-55-OOMD-testchill.service: Changed stop-sigterm -> failed
[   33.343653] TEST-55-OOMD.sh[826]: Jun 02 01:13:42 H systemd[1]: TEST-55-OOMD-testchill.service: Unit entered failed state.
[   33.347067] TEST-55-OOMD.sh[826]: Jun 02 01:13:42 H systemd[1]: TEST-55-OOMD-testchill.service: Releasing resources...
[   33.350897] TEST-55-OOMD.sh[592]: + exit 24
[   34.325740] reboot: Power down
Finishing after writing 63438 entries
Test failed, relevant logs can be viewed with: 

gh run download 9334201663 --name ci-mkosi-9334201663-1-fedora-40-failed-test-journals -D ci/ci-mkosi-9334201663-1-fedora-40-failed-test-journals && journalctl --file ci/ci-mkosi-9334201663-1-fedora-40-failed-test-journals/test/journal/TEST-55-OOMD.journal --no-hostname -o short-monotonic -u TEST-55-OOMD.service -p info
```

**Comment by ryantimwilson** [comment]
Did some debugging and found that it seems to be a regression in the `full` type in `memory.pressure` which is what systemd-oomd uses for oom kills

What we see on rawhide:
```
[root@H ~]# cat /sys/fs/cgroup/TEST.slice/TEST-55.slice/TEST-55-OOMD.slice/TEST-55-OOMD-workload.slice/TEST-55-OOMD-testmunch.service/memory.pressure
some avg10=79.51 avg60=23.28 avg300=5.32 total=17948445
full avg10=0.08 avg60=0.05 avg300=0.01 total=65046
```

What we usually see is some and full be equal for stress-ng workload:
```
[root@H ~]# cat /sys/fs/cgroup/TEST.slice/TEST-55.slice/TEST-55-OOMD.slice/TEST-55-OOMD-workload.slice/TEST-55-OOMD-testmunch.service/memory.pressure
some avg10=79.51 avg60=23.28 avg300=5.32 total=17948445
full avg10=79.51 avg60=23.28 avg300=5.32 total=17948445
```

Tried downgrading stress-ng but that didn't help.

Followed up with Johannes Weiner from the Linux kernel and here was his response:
```
There is a known scheduler bug in 6.12 that breaks psi. It leaks "running tasks" counts, which matches your symptoms of seeing partial pressure only.

Do you see "inconsistent task state" warnings in dmesg | grep psi?

A fix is queued in the scheduler tree, should be sent to Linus shortly: https://git.kernel.org/pub/scm/linux/kernel/git/tip/tip.git/commit/?id=c6508124193d42bbc3224571eb75bfa4c1821fbb
```

I did notice when running `journalctl` we see:
```
Oct 15 15:23:50 H kernel: psi: inconsistent task state! task=1:swapper/0 cpu=0 psi_flags=4 clear=0 set=4
```

Interestingly though...I don't see it for the stress-ng tasks even after several attempts. Kernel boot params indicate no rate limiting: systemd.log_ratelimit_kmsg=0 printk.devkmsg=on. I let Johannes know and am seeing what he comes back with. Might still be worth waiting for the fix though to see if that's the root cause.

**Comment by ryantimwilson** [comment]
Per Johannes:
```
Thanks for checking, this definitely looks like the issue I mentioned.

The state warning is issued only once per boot when a corruption is first detected to let you know the reporting will be unreliable going forward. You can't retrigger it. This particular bug tends to manifest very early during boot, so most likely by the time you checked it, it had scrolled out of the dmesg ring buffer. Luckily, it was captured by the journal.
```

This makes sense since the initial OOM kill tests seem to pass initially but later ones fail. So once the error happens, PSI reporting is completely unreliable.

So lets wait for the fix to get merged into 6.12

**Comment by bluca** [comment]
> So lets wait for the fix to get merged into 6.12

Thanks - could you please ask to cc the stable list as well, so that it gets picked up for 6.11.x too?

**Comment by ryantimwilson** [comment]
@bluca AFAICT the regression-causing commit - 152e11f6df29 ("sched/fair: Implement delayed dequeue") - isn't on 6.11 (e.g. I don't see SCHED_FEAT(DELAY_DEQUEUE, true) in sched/features.h). However, I can follow up to make sure if delayed dequeue gets pushed to 6.11, that we also get this patch.

**Comment by bluca** [comment]
Mmh but we see the test failures with kernel 6.11, so something must have happened there too?

**Comment by bluca** [comment]
No wait it seems it's 6.12 in Rawhide, never mind, ignore me then, sorry for the noise

**Comment by yuwata** [comment]
@ryantimwilson Please let us know when the kernel regression is fixed. We need to revert the workaround 88bbf187a9b2ebe0732caa1e886616ae5f8186da, waiting in #34833.

**Comment by ryantimwilson** [comment]
@yuwata looks like the kernel fix is in 6.12-rc4: https://github.com/torvalds/linux/commit/c6508124193d42bbc3224571eb75bfa4c1821fbb

Worth testing on Fedora rawhide again, not sure if Rawhide is pulling latest RC build for 6.12 or using static commit

**Comment by yuwata** [comment]
> @yuwata looks like the kernel fix is in 6.12-rc4: [torvalds/linux@c650812](https://github.com/torvalds/linux/commit/c6508124193d42bbc3224571eb75bfa4c1821fbb)
> 
> Worth testing on Fedora rawhide again, not sure if Rawhide is pulling latest RC build for 6.12 or using static commit

Thank you! It seems that Fedora finishes to build rc4, but it is not pushed to the repository yet. I will restart the test when it is available, hopefully within a few days.
